rfcs/41-ui-navigation.md

```diff
@@ -0,0 +1,859 @@
+  focused, it will act as a `NavRequest::Cancel` (leaving submenu to
+  enter the parent one).
+* **Lock**:
+  If we receive `NavRequest::Action` while this `Focusable` is
+  focused, the navigation system will freeze until `NavRequest::Free`
+  is received, sending a `NavEvent::Unlocked`.
+  \
+  This is useful to implement widgets with complex controls you don't
+  want to accidentally unfocus, or suspending the navigation system while
+  in-game.
+
+### `MenuSetting`
+
+See `bevy-ui-navigation` [`MenuSetting`] doc.
+
+The public API has `MenuSetting`, but we use internally [`TreeMenu`],
+this prevents end users from breaking assumptions about the menu trees.
+[More details on this decision](#spawning-menus).
+
+A menu that isolate children [`Focusable`]s from other
+focusables and specify navigation method within itself.
+
+A [`MenuSetting`] can be used to:
+* Prevent navigation from one specific submenu to another
+* Specify if 2d navigation wraps around the screen.
+* Specify "scope menus" such that a
+  `NavRequest::ScopeMove` emitted when
+  the focused element is a [`Focusable`] nested within this `MenuSetting`
+  will navigate this menu.
+* Specify _submenus_ and specify from where those submenus are reachable.
+* Specify which entity will be the parents of this [`MenuSetting`], see
+  `MenuSetting::reachable_from` or `MenuSetting::reachable_from_named` if you don't
+  have access to the `Entity`
+  for the parent [`Focusable`]
+
+If you want to specify which [`Focusable`] should be
+focused first when entering a menu, you should mark one of the children of
+this menu with `Focusable::prioritized`.
+
+#### Invariants
+
+**You need to follow those rules  to avoid panics**:
+1. A menu must have **at least one** [`Focusable`] child in the UI
+   hierarchy.
+2. There must not be a menu loop. I.e.: a way to go from menu A to menu B and
+   then from menu B to menu A while never going back.
+
+#### Panics
+
+Thankfully, programming errors are caught early and you'll probably get a
+panic fairly quickly if you don't follow the invariants.
+* Invariant (1) panics as soon as you add the menu without focusable
+  children.
+* Invariant (2) panics if the focus goes into a menu loop.
+
+#### Fields
+
+```rust
+pub struct MenuSetting {
+    /// Whether to wrap navigation.
+    ///
+    /// When the player moves to a direction where there aren't any focusables,
+    /// if this is true, the focus will "wrap" to the other direction of the screen.
+    pub wrapping: bool,
+    /// Whether this is a scope menu.
+    ///
+    /// A scope menu is controlled with [`NavRequest::ScopeMove`]
+    /// even when the focused element is not in this menu, but in a submenu
+    /// reachable from this one.
+    pub scope: bool,
+}
+```
+
+#### Going back to a previous menus
+
+`Focusable`s have a `prioritized` state that is set when they go from
+active/focused to not active.
+This is a form of memory that allows focus to go back to the last focused
```

Comment:
I think we need the ability for the developer to define: 

"When coming back to this menu, should we focus on the last-focused element or should we focus on the initial default element"

There are use cases in apps (filling in a new form) or games where we want to treat a menu as "first-time-user-visits" even though the user has visited before. The FFX battle system also springs to mind where you can choose between these two modes, which has a profound effect on muscle-memorisation.

Reply:
That's something I overlooked. Current design doesn't allow the user to modify the `prioritized` focusable. I think this can be enabled by providing a custom `SystemParam`.